### PR TITLE
Closes #10 add function `create_iso8601()`

### LINKED ISSUE #10: Develop a function to create ISO8601 date
### Feature Idea

**Purpose**
Converts collected date and optional time values into the iso8601 format. https://en.wikipedia.org/wiki/ISO_8601

**Functionality**
The feature should parse collected pieces of date and time data and reformat them into a valid iso8601 format. Partial dates are acceptable and common. Invalid dates can be encountered and should be handled as described in this specification.

EDC-collected dates generally arrive in multiple formats. Some examples are `dd MON yyyy`, `yyyymmdd`, `yyyy-mm-dd`

Times arrive in the `hh:mm` or `h:mm` (leading 0 implied) or `hh:mm:ss` or `hhmm` or `hhmmss` format 

We also need to handle data arriving as a single datetime field for example `dd MON yyy HH:MM:ss`

Format | Value | ISO 8601 Result
-- | -- | --
ddMMMyyyy | 12NOV2020 | 2020-11-12
dd MMM yyyy | 12 NOV 2020 | 2020-11-12
dd MMM yyyyhh:mm | 12 NOV 202015:15 | 2020-11-12T15:15
dd MMM yyyyhh:mm  | 03 NOV 20203:30 | 2020-11-03T03:30
dd MMM yyyyhh:mm  | UN DEC 201914:00 | 2019-12--T14:00
dd MMM yyyy  | 3 NOV 2020 | 2020-11-03
MMM yyyy  | NOV 2020 | 2020-11
MMM yyyy  | MAR 2019 | 2019-03
dd MMM yyyy  | UN APR 2018 | 2018-04
dd MMM yyyy  | UN UNK 2015 | 2015
MMM yyyy  | APR 2018 | 2018-04
dd MMM yyyy  | UK APR 2018 | 2018-04
dd MMM yyyy  | UN UNK 0 | Null because 0 is not a valid year
dd MMM yyyy  | UN UN 2015 | 2015
yyyymmdd | 20190212 | 2019-02-12
yyyymmddhh:mm | 2019020202:20 | 2019-02-02T02:20
yyyymmddhhmm | 2019-04-041045 | 2019-04-04T10:45
yyyymmddhhmm | 20200505null | 2020-05-05

### Relevant Input

one or many vectors and format for each vector.

### Relevant Output

A vector in ISO8601 date format.

### Reproducible Example/Pseudo Code

calculate_iso_date_time(vector1,format1...)

example
calculate_iso_date_time(ae_start_date,dd MON YYYY)

### PR DESCRIPTION
**Pull request by ramiromagno**
Thank you for your Pull Request! We have developed this task checklist from the
[Development Process
Guide](https://pharmaverse.github.io/admiraldev/devel/articles/development_process.html)
to help with the final steps of the process. Completing the below tasks helps to
ensure our reviewers can maximize their time on your code as well as making sure
the oak codebase remains robust and consistent.

Please check off each taskbox as an acknowledgment that you completed the task
or check off that it is not relevant to your Pull Request. This checklist is
part of the Github Action workflows and the Pull Request will not be merged into
the `devel` branch until you have checked off each task.

- [x] Place Closes #<insert_issue_number> into the beginning of your Pull z
Request Title (Use Edit button in top-right if you need to update)
- [x] Code is formatted according to the 
[tidyverse style guide](https://style.tidyverse.org/). Run 
`styler::style_file()` to style R and Rmd files
- [x] Updated relevant unit tests or have written new unit tests, which should
consider realistic data scenarios and edge cases, e.g. empty datasets, errors,
boundary cases etc. - See 
[Unit Test Guide](https://pharmaverse.github.io/admiraldev/devel/articles/unit_test_guidance.html#tests-should-be-robust-to-cover-realistic-data-scenarios)
- [x] Run `devtools::document()` so all `.Rd` files in the `man` folder and the
`NAMESPACE` file in the project root are updated appropriately
- [x] Address any updates needed for vignettes and/or templates
- [x] Update `NEWS.md` if the changes pertain to a user-facing function (i.e. it
has an `@export` tag) or documentation aimed at users (rather than developers)
- [x] Build oak site `pkgdown::build_site()` and check that all affected
examples are displayed correctly and that all new functions occur on the "[Reference](https://pharmaverse.github.io/admiral/cran-release/reference/index.html)" page. 
- [x] Address or fix all lintr warnings and errors - `lintr::lint_package()`
- [x] Run `R CMD check` locally and address all errors and warnings - `devtools::check()`

There is this NOTE: 

```
❯ checking package subdirectories ... NOTE
  Problems with news in ‘NEWS.md’:
  No news entries found.
```

- [x] Link the issue in the Development Section on the right hand side.
- [x] Address all merge conflicts and resolve appropriately

### CODE REVIEW

File: tests/testthat/test-create_iso8601.R
Comment (edgar-manukyan) [inline tests/testthat/test-create_iso8601.R:83]: Just to document that I tested with the Roche test cases and apart from three cases it is a match. I believe sdtm.oak behaves correctly.

```
# The only diffs I see are
# - "UN UNK 0" that roak make as NA, meanwhile sdtm.oak "----00"
# - "13NOV 2010", "33:10:12" that roak make as "2010-11-13", meanwhile sdtm.oak "2010-11-13T33:10:12
# - "2010.11.12", "33:93" that roak make as "2010-11-12", meanwhile sdtm.oak "-----T33:93"
```

```R
date_vector <- c(
  "12NOV2020", "12 NOV 2020", "12 NOV 2020", "03 NOV 2020",
  "UN DEC 2019", "3 NOV 2020", "NOV 2020", "MAR 2019", "UN APR 2018",
  "UN UNK 2015", "APR 2018", "UK APR 2018", "UN UNK 0", "UN UN 2015",
  "20190212", "20190202", "2019-04-04", "20200505", "33 MAR2020", "13NOV 2010",
  "UN NOV 2020", "UNNOV2020", "03UNK2020", "3NOV2020", "UNUNK2020", "UNUNK2020",
  "23.05.2019", "22-12-2010", "2010.11.12", "UN 1999", "UNK MAR 2020"
)

time_vector <- c(
  NA, NA, "15:15", "3:30", "14:00", NA, NA, NA, "NA", NA, NA,
  NA, NA, NA, NA, "02:20", "1045", "NA", NA, "33:10:12", "3:30", "3:18",
  "3:18", "0318", "0318", NA, "18:20", "1717", "33:93", NA, NA
)

date_formats <- dtc_formats |> 
  dplyr::filter(type %in% "date") |> 
  dplyr::pull(fmt) |> 
  c("dm y", "d.m.y", "y.m.d")

time_formats <- dtc_formats |> 
  dplyr::filter(type %in% "time") |> 
  dplyr::pull(fmt)

# 310K records
# user  system elapsed 
# 7.20    0.34    8.58
system.time({
  observed_output <- create_iso8601(
    date_vector,
    time_vector,
    .format = list(
      date_formats = date_formats, 
      time_formats = time_formats
    ),
    .na = c("UN", "UN UN", "UN UNK", "UNK", "UK"),
    .check_format = FALSE
  )
})

expected_output <- c(
  "2020-11-12", "2020-11-12", "2020-11-12T15:15",
  "2020-11-03T03:30", "2019-12--T14:00", "2020-11-03",
  "2020-11", "2019-03", "2018-04", "2015", "2018-04",
  "2018-04", "----00", "2015", "2019-02-12", "2019-02-02T02:20",
  "2019-04-04T10:45", "2020-05-05", NA, "2010-11-13T33:10:12",
  "2020-11--T03:30", "2020-11--T03:18", "2020---03T03:18",
  "2020-11-03T03:18", "2020----T03:18", "2020", "2019-05-23T18:20",
  "2010-12-22T17:17", "2010-11-12T33:93", "1999", "2020-03"
)
testthat::expect_identical(observed_output, expected_output)
```

**Review by edgar-manukyan — COMMENTED** [review]
Thanks so much @ramiromagno, this is a great job 💠 💠 💠 
So far I only had a chance to look at `create_iso8601()` and run the Roche tests. I added my comments along the lines.

I also did a time check with 310K records and it was slightly faster than the function we have in Roche:

```
# 310K records

# sdtm.oak::create_iso8601()
# user  system elapsed 
# 7.20    0.34    8.58

# roak::calc_iso_date_time()
# user  system elapsed 
# 7.51    0.79    9.22 
```

I will continue reviewing the internal functions 👍

File: tests/testthat/test-create_iso8601.R
Comment (edgar-manukyan) [reply tests/testthat/test-create_iso8601.R:83]: FYI @rammprasad

File: R/dtc_create_iso8601.R
Comment (ramiromagno) [reply R/dtc_create_iso8601.R:374]: Yes, absolutely. The documentation of `create_iso8601()` is a bit too short right now. It should definitely have more examples, particularly examples with these specials cases and examples with date and times being passed. Besides that we should have also a vignette or article, indeed. How are the guidelines right now? Do you make a distinction between a vignette and an article (only for pkgdown)?

File: R/dtc_create_iso8601.R
Comment (ramiromagno) [reply R/dtc_create_iso8601.R:321]: Yes, indeed!

File: R/dtc_create_iso8601.R
Comment (ramiromagno) [reply R/dtc_create_iso8601.R:328]: Okay, but note that here the `.na` parameter has a different meaning from your `replace_na` in roak. The `.na` parameter indicates which special values are allowed to be found while parsing without failing. So it does not mean the values with which you would replace missing values. In `create_iso8601()` missing values are always replaced with `"-"`.

File: tests/testthat/test-create_iso8601.R
Comment (ramiromagno) [reply tests/testthat/test-create_iso8601.R:83]: So in roak, we map `"33:10:12"` to NA/missing because of the 33, right? That's why at a certain point I asked whether to made semantic validation... I guess this is the case here? You find that 33 is nonsensical and invalidate the time, right? If you wish that can be easily introduced in sdtm.oak.

Comment (ramiromagno) [reply tests/testthat/test-create_iso8601.R:83]: In this example: `"UN UNK 0"` have you tried to supply `.na = c("UN", "UNK", "0")`?

Comment (ramiromagno) [reply tests/testthat/test-create_iso8601.R:83]:
```
create_iso8601("2010.11.12", "33:93", .format = c("y.m.d", "H:M"), .check_format = FALSE)
[1] "2010-11-12T33:93"
```

We have to be careful here with the dots `"."` or any character that might be interpreted as regex... This can be regarded as a strength or a bug... look at this example:

```
create_iso8601("2010X11Y12", "33:93", .format = c("y.m.d", "H:M"), .check_format = FALSE)
[1] "2010-11-12T33:93"
```
So actually the format `"y.m.d"` means an year, any character, month, any character, day.

Comment (edgar-manukyan) [reply tests/testthat/test-create_iso8601.R:83]: I also suggest that we mention in the `create_iso8601()`  that format is a regex and in the article provide some bad regexe examples as well to make them aware.

File: R/dtc_create_iso8601.R
Comment (edgar-manukyan) [reply R/dtc_create_iso8601.R:374]: In roak we were creating articles so far.

File: R/dtc_utils.R
Comment (edgar-manukyan) [inline R/dtc_utils.R:120]: Do we use `assert_capture_matrix()` anywhere?

File: R/package.R
Comment (edgar-manukyan) [inline R/package.R:1]: Do we use `dummy()` anywhere?

File: R/dtc_utils.R
Comment (edgar-manukyan) [inline R/dtc_utils.R:20]: Do we use `is_dtc_fmt` anywhere?

File: R/parse_dttm_fmt.R
Comment (edgar-manukyan) [inline R/parse_dttm_fmt.R:208]: Shall we declare all imports in the `package.R`?

File: R/parse_dttm_fmt.R
Comment (ramiromagno) [reply R/parse_dttm_fmt.R:161]: Yes, I am aware of it! I haven't yet made it a habit, though. But I will make those changes.

File: R/dtc_utils.R
Comment (ramiromagno) [reply R/dtc_utils.R:153]: Yes.

File: R/dtc_utils.R
Comment (ramiromagno) [reply R/dtc_utils.R:120]: We should. It was on my to-do list. Well spotted!

File: R/package.R
Comment (ramiromagno) [reply R/package.R:1]: No, I inherited that from Adam's commits and forgot to clean up. Will do.

File: R/dtc_utils.R
Comment (ramiromagno) [reply R/dtc_utils.R:20]: Not at the moment, but I should. I will use it such that it won't be an orphan function.

File: R/parse_dttm_fmt.R
Comment (ramiromagno) [reply R/parse_dttm_fmt.R:208]: Perhaps not all but definitely all those imports that are used very often, and I think rlang's `.data` certainly qualifies.

File: R/dtc_utils.R
Comment (ramiromagno) [reply R/dtc_utils.R:153]: @edgar-manukyan : When you mentioned data validation, were you implying that my tests on lines 80-83 were performing some sort of validation? Because I am not, right?

Comment (ramiromagno) [reply R/dtc_utils.R:187]: Resolved in https://github.com/pharmaverse/sdtm.oak/commit/673adccbd64b0b715697b4d77c9382cb0d57ba9b.

File: R/dtc_utils.R
Comment (ramiromagno) [reply R/dtc_utils.R:97]: Resolved in https://github.com/pharmaverse/sdtm.oak/commit/431c14a31f20ee707b5e5d4b2a3cbe4b3eb166c1.

File: R/dtc_utils.R
Comment (ramiromagno) [reply R/dtc_utils.R]: Ended up removing `is_dtc_fmt()`, see commit message for explanation: https://github.com/pharmaverse/sdtm.oak/commit/8cc595a81bb9c14625641bd54e769eccdcffbbcb.

File: R/parse_dttm_fmt.R
Comment (ramiromagno) [reply R/parse_dttm_fmt.R]: Done: https://github.com/pharmaverse/sdtm.oak/commit/ae8966c3800d154c4b4c5c1f9ace7ae47a3a4713.

File: R/package.R
Comment (ramiromagno) [reply R/package.R:1]: Done: https://github.com/pharmaverse/sdtm.oak/commit/00d1e8feda72aecddffc2057a682a906e949b98a.

File: R/dtc_utils.R
Comment (ramiromagno) [reply R/dtc_utils.R:139]: Done: https://github.com/pharmaverse/sdtm.oak/commit/431c14a31f20ee707b5e5d4b2a3cbe4b3eb166c1.

File: R/dtc_formats.R
Comment (galachad) [inline R/dtc_formats.R:14]: This is a super small dataset. I wonder why you prefer to keep that as generated rd file over just code?

File: R/dtc_utils.R
Comment (galachad) [inline R/dtc_utils.R:19]: We do need `assert_character_vectora` as it should be covered by `rlang::arg_match`. Is there any reason for using it here?

File: NAMESPACE
Comment (galachad) [inline NAMESPACE:3]: So we decided to use `rlang::`. I was pushing a little bit to import rlang into package, but both approach are good for me, so we should use `rlang::`

**Review by galachad — COMMENTED** [review]
Few small comments from me, please apply the styler changes into the code.

File: R/dtc_formats.R
Comment (ramiromagno) [reply R/dtc_formats.R:14]: Hi @galachad: Indeed the dataset is small right now but likely to grow as we support more formats... I am not sure I understood your point _keep that as generated rd file over just code_? Do you mean keeping the `dtc_formats` as an internal R object only and not export it? The reason for exporting it and to document it is that the users need to find somewhere in the documentation which formats are allowed, and this dataset (and its documentation) provide that.

File: R/sdtm.oak-package.R
Comment (ramiromagno) [reply R/sdtm.oak-package.R:6]: We need `#' @importFrom tibble tibble` if we want tibbles to benefit from methods for generics like `print()` or `[`. If we stick with the exported `dtc_formats` below then I find this a nice feature. The `.data` pronoun is used in R/dtc_utils.R and R/parse_dttm_fmt.R.

File: R/dtc_utils.R
Comment (ramiromagno) [reply R/dtc_utils.R:19]: There is no need indeed, I will remove it! Thanks for spotting this.

Comment (ramiromagno) [reply R/dtc_utils.R]: Resolved in https://github.com/pharmaverse/sdtm.oak/pull/21/commits/d6cd751dfd06d9747cfa528974c7ae449b381d01.

**Review by galachad — APPROVED** [review]

**Comment by edgar-manukyan** [comment]
@ramiromagno, thanks so much for addressing Adam's and my comments 🙏 

Everything looks awesome! I know that after the second approval you can merge, just wanted to make sure that at lease one of our subject matter experts had a chance to try this awesomeness out. @rammprasad, will you have time?

**Review by edgar-manukyan — APPROVED** [review]

**Comment by rammprasad** [comment]
Yes, I am reviewing it today. Lets merge it tomorrow.

**Comment by ramiromagno** [comment]
Thank you so much for your feedback so far. I think we can still improve quite a bit the article about the `create_iso8601()` function and add more units tests to get a good coverage. But perhaps getting some user feedback already would be useful too. Note that, as @rammprasad had suggested in one of our meetings, I changed the default value of `.check_format` from `TRUE` to `FALSE`.

File: R/dtc_create_iso8601.R
Comment (rammprasad) [inline R/dtc_create_iso8601.R:329]: Do we need to add an example on how to use fmt_cmp()

File: data-raw/dtc_formats.R
Comment (rammprasad) [inline data-raw/dtc_formats.R:16]: Can we also add the below date and time formats? `-` indicates an unknown component in it.

dd MMM yyyy HH nn ss
HH nn
HH:nn:ss
MMM-
HH nn ss
dd MMM yyyy HH mm ss
dd- MMM- yyyy
dd MMM yyyy HH:nn:ss

HH:nn
dd MMM yyyy hh:nn:ss:rr
dd- MMM- yyyy-
dd MMM yyyy hh nn ss
dd MMM yyyy hh:nn:ss
MMM- yyyy
dd- MMM- yyyy hh:mm:ss
dd- MMM- yyyy HH nn ss
dd- MMM- yyyy- HH nn
dd- MMM yyyy
dd MMM yyyy
dd MMM yyyy HH:nn
hh:nn
yy:mm
yyyy
nn ss
HH mm
dd MMM yyyy hh ss
dd MMM yyyy HH nn
MMM yyyy
HH+ nn

if we do not want to add them, we need to find a way for this function to process it. Currently, I am not able to provide dd MMM yyyy, it gives me an error.

File: R/dtc_create_iso8601.R
Comment (rammprasad) [inline R/dtc_create_iso8601.R:414]: Didn't we agree to pass a warning message if some values do not adhere to the specified format?

create_iso8601(c("UN UNK 201914:00", "UN JAN 2021"), .format = "dd mmm yyyy", .check_format = FALSE, .na = c("UN", "UNK"))

[1] NA        "2021-01"

The warning message can read as "Some source values does not adhere to the format provided, returning null values"

Maybe we need to create an exception listing. It can be developed as a separate issue.

File: R/dtc_create_iso8601.R
Comment (rammprasad) [inline R/dtc_create_iso8601.R:405]: This doesn't work in debug mode when I give 

dots <- c("UN UNK 201914:00", "UN JAN 2021")
.format = "dd mmm yyyy"

**Review by rammprasad — COMMENTED** [review]
Great work, @ramiromagno . it is actually mindblowing. I have added some comments. Please take a look.

File: data-raw/dtc_formats.R
Comment (ramiromagno) [reply data-raw/dtc_formats.R:16]: So `"MMM-"` means month, including unknown values? But what forms can these unknown values take?

File: R/dtc_create_iso8601.R
Comment (ramiromagno) [reply R/dtc_create_iso8601.R:414]: Yes, we did agree on this. But like you said, I think it is best provided as a separate issue.

File: R/dtc_create_iso8601.R
Comment (ramiromagno) [reply R/dtc_create_iso8601.R:405]: You would have to do it like this:

``` r
library(sdtm.oak)
datetimes <- c("UN UNK 201914:00", "UN JAN 2021")
create_iso8601(datetimes, .format = list(c("dd mmm yyyy", "dd mmm yyyyHH:MM")), .na = c("UN", "UNK"))
#> [1] "2019----T14:00" "2021-01"
```

<sup>Created on 2023-11-22 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

File: data-raw/dtc_formats.R
Comment (ramiromagno) [reply data-raw/dtc_formats.R:16]: The `"rr"` in `"dd MMM yyyy hh:nn:ss:rr"` was not in the original specification request as this is an additional time component beyond hours, minutes and seconds. Is this meant to be fractional seconds?

File: R/dtc_create_iso8601.R
Comment (ramiromagno) [reply R/dtc_create_iso8601.R:394]: Hi @rammprasad: I am not sure which lower-level and wrapper functions you are referring to. If by lower-level function you mean the `create_iso8601()` function, and some other derivation-function as the wrapper, then the try/catch is to be implemented at the wrapper function body. If that is the case, then there is nothing to be done on my end here. Otherwise, if you are suggesting to add the try/catch block around the `rlang::abort()` command, then we are better off not aborting in the first place and decide on a reasonable return value for `create_iso8601()` when this condition happens.

File: R/dtc_create_iso8601.R
Comment (ramiromagno) [reply R/dtc_create_iso8601.R:329]: Done: https://github.com/pharmaverse/sdtm.oak/pull/21/commits/aa2113556c29395161fddae73eea72e3dbe5ba9b.

File: R/dtc_create_iso8601.R
Comment (ramiromagno) [reply R/dtc_create_iso8601.R:421]: Example added: https://github.com/pharmaverse/sdtm.oak/pull/21/commits/81336b18f5fdd41acc93c1a657bd0ea95ad765d8.

File: R/dtc_create_iso8601.R
Comment (ramiromagno) [reply R/dtc_create_iso8601.R:429]: For the warnings' feature and feedback on which values failed, let us track the progress here: https://github.com/pharmaverse/sdtm.oak/issues/29.

File: R/dtc_create_iso8601.R
Comment (ramiromagno) [reply R/dtc_create_iso8601.R:409]: This issue about the try/catch block is to be addressed in the appropriate wrapper function, so I am closing this conversation.

**Comment by ramiromagno** [comment]
Hello @rammprasad: Just to be sure, I am only waiting for this pull request to be reviewed by you and then merged before I tackle these other two issues: #28 and #29. Is that okay?

**Comment by rammprasad** [comment]
Thank you, @ramiromagno, for merging this to the main.  We can work on other issues (maybe #18 ) before #28 and #29

**Pull request merged**
